File: sdks/python-sdk/quantaureum/utils/hex.py
# Quantaureum Python SDK source, version 1.0.0.
"""Hexadecimal conversion utilities for the SDK.

Property 5: Hex Conversion Round-Trip
For any bytes object, from_hex(to_hex(data)) should equal the original data.
Validates: Requirements 5.5
"""

from typing import Union

from ..exceptions import ValidationError


def to_hex(value: Union[int, bytes]) -> str:
    """Convert an integer or bytes to a hex string with 0x prefix.

    Args:
        value: An integer or bytes object to convert.

    Returns:
        A hex string with 0x prefix.

    Raises:
        ValidationError: If the value is not an int or bytes.

    Examples:
        >>> to_hex(255)
        '0xff'
        >>> to_hex(b'\\x00\\x01\\x02')
        '0x000102'
    """
    if isinstance(value, int):
        if value < 0:
            raise ValidationError(
                message="Cannot convert negative integer to hex",
                field="value",
                value=value,
                expected="non-negative integer",
            )
        # Handle zero case
        if value == 0:
            return "0x0"
        return hex(value)
    elif isinstance(value, bytes):
        return "0x" + value.hex()
    else:
        raise ValidationError(
            message=f"Cannot convert {type(value).__name__} to hex",
            field="value",
            value=value,
            expected="int or bytes",
        )


def from_hex(hex_str: str) -> bytes:
    """Convert a hex string to bytes.

    Args:
        hex_str: A hex string, optionally with 0x prefix.

    Returns:
        The decoded bytes.

    Raises:
        ValidationError: If the hex string is invalid.

    Examples:
        >>> from_hex('0x000102')
        b'\\x00\\x01\\x02'
        >>> from_hex('ff')
        b'\\xff'
    """
    if not isinstance(hex_str, str):
        raise ValidationError(
            message=f"Expected string, got {type(hex_str).__name__}",
            field="hex_str",
            value=hex_str,
            expected="string",
        )

    # Remove 0x prefix if present
    if hex_str.startswith(("0x", "0X")):
        hex_str = hex_str[2:]

    # Handle empty string
    if not hex_str:
        return b""

    # Pad with leading zero if odd length
    if len(hex_str) % 2 != 0:
        hex_str = "0" + hex_str

    try:
        return bytes.fromhex(hex_str)
    except ValueError as e:
        raise ValidationError(
            message=f"Invalid hex string: {e}",
            field="hex_str",
            value=hex_str,
            expected="valid hex characters (0-9, a-f, A-F)",
        ) from e


def to_hex_with_size(value: int, byte_size: int) -> str:
    """Convert an integer to a hex string with a specific byte size.

    Args:
        value: An integer to convert.
        byte_size: The number of bytes in the output.

    Returns:
        A hex string with 0x prefix, zero-padded to the specified size.

    Raises:
        ValidationError: If the value doesn't fit in the specified size.

    Examples:
        >>> to_hex_with_size(255, 2)
        '0x00ff'
        >>> to_hex_with_size(1, 32)
        '0x0000000000000000000000000000000000000000000000000000000000000001'
    """
    if not isinstance(value, int):
        raise ValidationError(
            message=f"Expected int, got {type(value).__name__}",
            field="value",
            value=value,
            expected="int",
        )

    if value < 0:
        raise ValidationError(
            message="Cannot convert negative integer to hex",
            field="value",
            value=value,
            expected="non-negative integer",
        )

    max_value = (1 << (byte_size * 8)) - 1
    if value > max_value:
        raise ValidationError(
            message=f"Value {value} exceeds maximum for {byte_size} bytes",
            field="value",
            value=value,
            expected=f"value <= {max_value}",
        )

    hex_chars = byte_size * 2
    return "0x" + format(value, f"0{hex_chars}x")


def is_hex_string(value: str) -> bool:
    """Check if a string is a valid hex string.

    Args:
        value: The string to check.

    Returns:
        True if the string is a valid hex string, False otherwise.

    Examples:
        >>> is_hex_string('0x1234')
        True
        >>> is_hex_string('1234')
        True
        >>> is_hex_string('0xGHIJ')
        False
    """
    if not isinstance(value, str):
        return False

    # Remove 0x prefix if present
    if value.startswith(("0x", "0X")):
        value = value[2:]

    if not value:
        return True  # Empty hex string is valid

    # Check all characters are valid hex
    try:
        int(value, 16)
        return True
    except ValueError:
        return False


__all__ = ["to_hex", "from_hex", "to_hex_with_size", "is_hex_string"]

File: sdks/ts-sdk/src/utils/hex.ts
// Quantaureum TypeScript SDK source, version 1.0.0.
/**
 * Hex conversion utilities for Quantaureum SDK
 * @module utils/hex
 */

import { ValidationError, ErrorCode } from '../errors';

/**
 * Convert a value to a hex string with 0x prefix
 * @param value - Number, bigint, or Uint8Array to convert
 * @returns Hex string with 0x prefix
 */
export function toHex(value: number | bigint | Uint8Array): string {
  if (value instanceof Uint8Array) {
    return hexlify(value);
  }

  if (typeof value === 'bigint') {
    if (value < 0n) {
      throw new ValidationError(
        'Cannot convert negative bigint to hex',
        'value',
        value,
        ErrorCode.INVALID_ARGUMENT,
        'non-negative bigint'
      );
    }
    const hex = value.toString(16);
    return '0x' + hex;
  }

  if (typeof value === 'number') {
    if (!Number.isInteger(value) || value < 0) {
      throw new ValidationError(
        'Cannot convert non-integer or negative number to hex',
        'value',
        value,
        ErrorCode.INVALID_ARGUMENT,
        'non-negative integer'
      );
    }
    return '0x' + value.toString(16);
  }

  throw new ValidationError(
    'Invalid value type for hex conversion',
    'value',
    value,
    ErrorCode.INVALID_ARGUMENT,
    'number, bigint, or Uint8Array'
  );
}

/**
 * Convert a hex string to Uint8Array
 * @param hex - Hex string (with or without 0x prefix)
 * @returns Uint8Array of bytes
 */
export function fromHex(hex: string): Uint8Array {
  if (typeof hex !== 'string') {
    throw new ValidationError(
      'Expected hex string',
      'hex',
      hex,
      ErrorCode.INVALID_HEX,
      'string'
    );
  }

  // Remove 0x prefix if present
  let cleanHex = hex.startsWith('0x') || hex.startsWith('0X') ? hex.slice(2) : hex;

  // Validate hex characters
  if (!/^[0-9a-fA-F]*$/.test(cleanHex)) {
    throw new ValidationError(
      'Invalid hex string: contains non-hex characters',
      'hex',
      hex,
      ErrorCode.INVALID_HEX,
      'valid hex characters (0-9, a-f, A-F)'
    );
  }

  // Pad with leading zero if odd length
  if (cleanHex.length % 2 !== 0) {
    cleanHex = '0' + cleanHex;
  }

  const bytes = new Uint8Array(cleanHex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(cleanHex.slice(i * 2, i * 2 + 2), 16);
  }

  return bytes;
}

/**
 * Convert Uint8Array to hex string with 0x prefix
 * @param data - Uint8Array to convert
 * @returns Hex string with 0x prefix
 */
export function hexlify(data: Uint8Array): string {
  if (!(data instanceof Uint8Array)) {
    throw new ValidationError(
      'Expected Uint8Array',
      'data',
      data,
      ErrorCode.INVALID_ARGUMENT,
      'Uint8Array'
    );
  }

  let hex = '0x';
  for (let i = 0; i < data.length; i++) {
    const byte = data[i];
    if (byte !== undefined) {
      hex += byte.toString(16).padStart(2, '0');
    }
  }
  return hex;
}

/**
 * Convert hex string to Uint8Array (alias for fromHex)
 * @param hex - Hex string (with or without 0x prefix)
 * @returns Uint8Array of bytes
 */
export function arrayify(hex: string): Uint8Array {
  return fromHex(hex);
}

/**
 * Check if a string is a valid hex string
 * @param value - String to check
 * @returns True if valid hex string
 */
export function isHexString(value: unknown): value is string {
  if (typeof value !== 'string') {
    return false;
  }

  const cleanHex = value.startsWith('0x') || value.startsWith('0X') ? value.slice(2) : value;
  return /^[0-9a-fA-F]*$/.test(cleanHex);
}

/**
 * Pad a hex string to a specific byte length
 * @param hex - Hex string to pad
 * @param length - Target byte length
 * @returns Padded hex string with 0x prefix
 */
export function zeroPadHex(hex: string, length: number): string {
  if (!isHexString(hex)) {
    throw new ValidationError(
      'Invalid hex string',
      'hex',
      hex,
      ErrorCode.INVALID_HEX,
      'valid hex string'
    );
  }

  const cleanHex = hex.startsWith('0x') || hex.startsWith('0X') ? hex.slice(2) : hex;
  const targetLength = length * 2; // Each byte is 2 hex chars

  if (cleanHex.length > targetLength) {
    throw new ValidationError(
      `Hex string too long: ${cleanHex.length / 2} bytes, max ${length} bytes`,
      'hex',
      hex,
      ErrorCode.INVALID_ARGUMENT,
      `hex string with at most ${length} bytes`
    );
  }

  return '0x' + cleanHex.padStart(targetLength, '0');
}

/**
 * Concatenate multiple hex strings
 * @param hexStrings - Hex strings to concatenate
 * @returns Concatenated hex string with 0x prefix
 */
export function concatHex(...hexStrings: string[]): string {
  let result = '0x';
  for (const hex of hexStrings) {
    if (!isHexString(hex)) {
      throw new ValidationError(
        'Invalid hex string in concatenation',
        'hexStrings',
        hex,
        ErrorCode.INVALID_HEX,
        'valid hex string'
      );
    }
    const cleanHex = hex.startsWith('0x') || hex.startsWith('0X') ? hex.slice(2) : hex;
    result += cleanHex;
  }
  return result;
}
